elements/pl-big-o-input/big_o_utils.py

```diff
@@ -0,0 +1,190 @@
+from typing import Callable, List, Tuple
+import sympy
+import python_helper_sympy as phs
+
+BigOGradingFunctionT = Callable[[str, str, List[str]], Tuple[float, str]]
+
+CORRECT_UNCONDITIONAL_FEEDBACK = "Correct!"
+CORRECT_COMPLEX_FEEDBACK = (
+    "Correct! Note that your expression may be unnecessarily complex."
+)
+NEGATIVE_FEEDBACK = "Your expression is negative."
+INCORRECT_FEEDBACK = "Your answer is incorrect."
+TOO_LOOSE_FEEDBACK = "Your answer is correct, but too loose."
+LOWER_ORDER_TERMS_FEEDBACK = (
+    "Your answer is correct, but you have unnecessary lower order terms."
+)
+CONSTANT_FACTORS_FEEDBACK = (
+    "Your answer is correct but has unncessary constant factors."
+)
+
+THETA_CONSTANT_FACTORS_FEEDBACK = (
+    "Incorrect, your answer has unnecessary constant factors."
+)
+THETA_LOWER_ORDER_TERMS_FEEDBACK = (
+    "Incorrect, your answer has unnecessary lower order terms."
+)
+
+
+def grade_big_o_expression(
+    a_true: str, a_sub: str, variables: List[str]
+) -> Tuple[float, str]:
+    if a_true == a_sub:
+        return (1, CORRECT_UNCONDITIONAL_FEEDBACK)
+
+    sym_true = phs.convert_string_to_sympy(
+        a_true, variables, allow_complex=False, allow_trig_functions=False
+    )
+    sym_sub = phs.convert_string_to_sympy(
+        a_sub, variables, allow_complex=False, allow_trig_functions=False
+    )
+
+    if sym_true.equals(sym_sub):
+        return (1, CORRECT_COMPLEX_FEEDBACK)
```

Comment:
Currently yes. Looking at this again, we can actually check for structural equality above instead of string equality (since structural doesn't evaluate expressions like `1+1` as equal to `2`). Will change.

Reply:
If we can check for structural equality, is there a reason we wouldn't just do that instead of checking for string equality?

Reply:
So the issue is that structural equality simplifies literals. For example, if you input `(1+1)**n` and the reference solution is `2**n`, these would be treated as equal under structural equality using `==`. We haven't found a way of comparison that avoids this simplification.

However, if we find something that works in this case, it's very straightforward to change the comparison to this and add tests.

Reply:
Ah, ok - that's unfortunate. I've spent approximately zero time checking how feasible this is, but I imagine you could improve over simple string equality by walking the AST and comparing nodes, which would in theory avoid the problem of `n**2` and `n**(2)` not being equivalent. Using `parse_expr('..', evaluate=False)` and a traversal like the one described in https://stackoverflow.com/questions/59816730/run-through-the-ast-of-an-expression-in-sympy might get you pretty close. (Not requesting that you implement here, just food for thought.)

Reply:
I think that sounds reasonable as a feature to add later on. Right now, most runtimes are only a single term (something like `n**3`), so removing just spaces gets us pretty close already.

Reply:
Right, makes sense! As long as students get full credit for a correct answer (which they currently do), that's all that really matters 🙂